ingestion-beam/src/main/java/com/mozilla/telemetry/ContextualServicesReporter.java

```diff
@@ -70,31 +72,65 @@ public static PipelineResult run(ContextualServicesReporterOptions.Parsed option
         .failuresTo(errorCollections) //
         .apply(EmitCounters.of());
 
-    Set<String> aggregatedDocTypes = ImmutableSet.of("topsites-impression");
-    Set<String> perContextIdDocTypes = ImmutableSet.of("topsites-click");
+    Set<String> individualImpressions = ImmutableSet.of("topsites-impression");
+    Set<String> individualClicks = ImmutableSet.of("topsites-click");
+
     Set<String> unionedDocTypes = Stream
-        .concat(aggregatedDocTypes.stream(), perContextIdDocTypes.stream())
+        .concat(individualImpressions.stream(), individualClicks.stream())
         .collect(Collectors.toSet());
 
+    // Perform windowed click counting per context_id, adding a click-status to the reporting URL
+    // if the count passes a threshold.
+    PCollection<SponsoredInteraction> clicksCountedByContextId = requests
+        .apply("FilterPerContextIdDocTypes", Filter.by((interaction) -> individualClicks //
+            .contains(interaction.getDerivedDocumentType())))
+        .apply(LabelSpikes.perContextId(options.getClickSpikeThreshold(),
+            Time.parseDuration(options.getClickSpikeWindowDuration()), TelemetryEventType.CLICK));
+
+    // Perform windowed impression counting per context_id, adding an impression-status to the
+    // reporting URL
+    // if the count passes a threshold.
+    PCollection<SponsoredInteraction> impressionsCountedByContextId = requests
+        .apply("FilterPerContextIdDocTypes", Filter.by((interaction) -> individualImpressions //
+            .contains(interaction.getDerivedDocumentType())))
+        .apply(LabelSpikes.perContextId(options.getImpressionSpikeThreshold(),
+            Time.parseDuration(options.getImpressionSpikeWindowDuration()),
+            TelemetryEventType.IMPRESSION));
+
     // Aggregate impressions.
-    PCollection<SponsoredInteraction> aggregated = requests
-        .apply("FilterAggregatedDocTypes", Filter.by((interaction) -> aggregatedDocTypes //
+    PCollection<SponsoredInteraction> aggregatedGenuineImpressions = requests
+        .apply("FilterAggregatedDocTypes", Filter.by((interaction) -> individualImpressions //
             .contains(interaction.getDerivedDocumentType())))
+        .apply("FilterForLegitImpressions",
+            Filter.by(new SerializableFunction<SponsoredInteraction, Boolean>() {
+
+              @Override
+              public Boolean apply(SponsoredInteraction input) {
+                return !input.getReportingUrl().contains("impression-status=1");
+              }
+            }))
```

Comment:
I don't think we need to split impression aggregation by `impression-status` looking through `AggregateImpressions::getAggregationKey` it looks like it sorts all the url parameters and uses the full URL for the aggregation key.

Reply:
@quiiver changes made! Running new job now

Reply:
@quiiver build on the way!

https://console.cloud.google.com/dataflow/jobs/us-central1/2023-03-08_13_19_12-18184766585711042219;graphView=0?project=contextual-services-dev&authuser=1&pageState=(%22dfTime%22:(%22l%22:%22dfJobMaxTime%22))